### ISSUE TITLE
Models with the same objects in more than one collection render incorrectly

### ISSUE THREAD

**Issue by jksaunders** [grails/grails-views#175]
I've pared down my issue to a project with two domain classes: `Bulletin`s and `User`s, where each `Bulletin` has `content` and a collection of `User`s called `targetUsers` and another collection of `User`s called `contactUsers`.

I'm experiencing a painful issue where when rendering a model of a `Bulletin` that has the same users in both `targetUsers` and `contactUsers`, values defined in the closure get completely mixed up. Values given with `includes` work however.

In the attached example, there are 3 users: Anthony, Jordyn and Josh with ids 1, 2 and 3 respectively. There is one `Bulletin` that has all 3 users as `contactUsers` and `targetUsers`.

`_bulletin.gson`:
```
import com.jksaunders.Bulletin

model {
    Bulletin bulletin
}

json g.render(bulletin, [includes: ['content']]) {
    contactUsers tmpl.'/user/userMinimumInfo'('user', bulletin.contactUsers ?: [])
    targetUsers tmpl.'/user/userMinimumInfo'('user', bulletin.targetUsers ?: [])
}
```

`userMinimumInfo.gson`:
```
import com.jksaunders.User

model {
    User user
}

json g.render(user, [includes: ['id', 'name']]) {
    username user.username
}
```

The above results in the following output:
```
{
  "content": "Hi everyone!",
  "contactUsers": [
    {
      "id": 2,
      "name": "",
      "username": "Jordyn"
    },
    {
      "id": 1,
      "name": "",
      "username": "Anthony"
    },
    {
      "id": 3,
      "name": "",
      "username": "Josh"
    }
  ],
  "targetUsers": [
    {
      "id": 2,
      "name": "",
      "username": "Josh" <------ incorrect!
    },
    {
      "id": 1,
      "name": "",
      "username": "Anthony"
    },
    {
      "id": 3,
      "name": "",
      "username": "Josh"
    }
  ]
}
```

You can see that the `username` value for `User` id 2 has gotten messed up. Again, it can be noted that the values specified by `includes` continue to work consistently (i.e. the proper `id` is shown). If I use `includes` to get the `username` in the above example instead of in the closure, it works as expected.

In my full project, the problem is exacerbated because I specify many of my values in the closure rather than using `includes` or letting the entire object be rendered. The main use case for using the closure is to rename some of the field values (eg. in the JSON output, show the `publicId` value using the key `id` instead).

This happens consistently and can even be replicated by changing `_bulletin.gson` to the following:
```
import com.jksaunders.Bulletin

model {
    Bulletin bulletin
}

json g.render(bulletin, [includes: ['content']]) {
    targetUsers tmpl.'/user/userMinimumInfo'('user', bulletin.targetUsers ?: [])
    targetUsers2 tmpl.'/user/userMinimumInfo'('user', bulletin.targetUsers ?: [])
}
```
Even just rendering the same collection twice results in different output - the first one being correct and the second one always being incorrect.

I tried to follow along using breakpoints to see where the issue was happening and it looks like in some cases an object is considered to have been already processed, but when it's given a value, it's given the value of the wrong processed object. I'd be happy to help resolve it if pointed in the right direction! I'd also be happy to hear that I'm using the tool incorrectly. Thanks in advance 🙏 

In the attached example’s `Bootstrap.groovy`, I've specified the 3 `User`s and the single `Bulletin`, and you can navigate to `ApiBulletinController` and see the JSON output rendered in the `index` method (`http://localhost:8080/apiBulletin/index`).

[grails-issues-2.zip](https://github.com/grails/grails-views/files/2173339/grails-issues-2.zip)

**Comment by ilopmar** [comment]
I've discovered where it's failing and I have a workaround for you. The issue is in `_userMinimumInfo.gson` template. Your original version is:
```groovy
model {
    User user
}

json g.render(user, [includes: ['id', 'name']]) {
    username user.username
}
```

If you change it to:

```groovy
model {
    User user
}

json {
    id user.id
    name user.name
    username user.username
}
```

Everything works as expected. I'll work on a fix for this and it will be included in the next `1.2.9` version.

Btw, an additional comment about your project. You're using an old and not final Grails version `3.3.0.RC1`. I would recommend you to upgrade to the latest `3.3.6` and also upgrade the dependencies: gorm,... and specially json-views which is now `1.2.8`.

**Comment by jksaunders** [comment]
Thanks so much for the fix and workaround! Much appreciated. Thanks as well for the tip on upgrading - I tried upgrading to see if it would help with this issue but yeah I definitely need to set aside time to bring it all up to date.
